base/zfp_compressor_body.hpp

```diff
@@ -2,10 +2,82 @@
 
 #include "base/zfp_compressor.hpp"
 
+#include <functional>
+#include <vector>
+#include <string>
+
 namespace principia {
 namespace base {
 namespace zfp_compressor_internal {
 
+template<int D>
+class NDimensionalHelper {
+ public:
+  using unique_zfp_field =
+      std::unique_ptr<zfp_field, std::function<void(zfp_field*)>>;
+
+  static unique_zfp_field NewField(zfp_type type, std::vector<double>& v);
+
+ private:
+  // Returns the size rounded up to a multiple of 4^(D - 1).
+  static constexpr std::int64_t RoundUp(std::int64_t size);
+
+  // The ZFP block size.
+  static constexpr int block_ = 4;
+  // Data must be padded to a multiple of that value, which is 4^(D - 1).
+  static constexpr int padding_ = 1 << (2 * (D - 1));
+};
+
+template<int D>
+typename NDimensionalHelper<D>::unique_zfp_field
+NDimensionalHelper<D>::NewField(zfp_type const type, std::vector<double>& v) {
+  auto free = [](zfp_field* const field) { zfp_field_free(field); };
+
+  // On compression: Round up the size of the vector to a multiple of the block
+  // size.  This will lead to poor compression at the end, but there is no
+  // support for "ignored" data in zfp at this point.
+  // On decompression: Make sure that we have enough space in the vector to
+  // decompress the padding.
+  v.resize(RoundUp(v.size()), 0);
+  CHECK_EQ(0, v.size() % padding_);
+
+  // Beware nx, ny and friends!  (And the Jabberwock, my son!)
+  // See
+  // https://zfp.readthedocs.io/en/release0.5.5/tutorial.html#high-level-c-interface
+  if constexpr (D == 1) {
+    return unique_zfp_field(
+        zfp_field_1d(v.data(), type, /*nx=*/v.size() / padding_),
+        std::move(free));
+  } else if constexpr (D == 2) {
+    return unique_zfp_field(
+        zfp_field_2d(v.data(), type, /*nx=*/block_, /*ny=*/v.size() / padding_),
+        std::move(free));
+  } else if constexpr (D == 3) {
+    return unique_zfp_field(zfp_field_3d(v.data(),
+                                         type,
+                                         /*nx=*/block_,
+                                         /*ny=*/block_,
+                                         /*nz=*/v.size() / padding_),
+                            std::move(free));
+  } else if constexpr (D == 4) {
+    return unique_zfp_field(zfp_field_4d(v.data(),
+                                         type,
+                                         /*nx=*/block_,
+                                         /*ny=*/block_,
+                                         /*nz=*/block_,
+                                         /*nw=*/v.size() / padding_),
+                            std::move(free));
+  } else {
+    static_assert(false, "Unsupported dimension");
+  }
+}
+
+template<int D>
+constexpr std::int64_t NDimensionalHelper<D>::RoundUp(std::int64_t const size) {
+  // Hacker's Delight, H. S. Warren, Jr., section 3-1.
```

Comment:
Add this to the bibliography, refer to the generated abbreviation.

Reply:
Done.